tls/s2n_ktls.c

```diff
@@ -0,0 +1,199 @@
+     * - Confirm application is not using custom I/O
+     * - Confirm kTLS isn't enabled already
+     */
+    switch (ktls_mode) {
+        case S2N_KTLS_MODE_SEND:
+            RESULT_ENSURE_REF(conn->send_io_context);
+            RESULT_ENSURE(!conn->managed_send_io, S2N_ERR_KTLS);
+            RESULT_ENSURE(!conn->ktls_send_enabled, S2N_ERR_KTLS);
+            break;
+        case S2N_KTLS_MODE_RECV:
+            RESULT_ENSURE_REF(conn->recv_io_context);
+            RESULT_ENSURE(!conn->managed_recv_io, S2N_ERR_KTLS);
+            RESULT_ENSURE(!conn->ktls_recv_enabled, S2N_ERR_KTLS);
+            break;
+        case S2N_KTLS_MODE_DISABLED:
+        case S2N_KTLS_MODE_DUPLEX:
+            RESULT_BAIL(S2N_ERR_SAFETY);
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd)
+{
+    RESULT_GUARD(s2n_ktls_validate_socket_mode(conn, ktls_mode));
+    RESULT_ENSURE_REF(fd);
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV) {
+        /* retrieve the receive fd */
+        const struct s2n_socket_read_io_context *peer_socket_ctx = conn->recv_io_context;
+        *fd = peer_socket_ctx->fd;
+    } else if (ktls_mode == S2N_KTLS_MODE_SEND) {
+        /* retrieve the send fd */
+        const struct s2n_socket_write_io_context *peer_socket_ctx = conn->send_io_context;
+        *fd = peer_socket_ctx->fd;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+#ifdef S2N_PLATFORM_SUPPORTS_KTLS
+
+S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_GUARD(s2n_ktls_validate_socket_mode(conn, ktls_mode));
+
+    /* register the tls ULP */
+    int fd = 0;
+    RESULT_GUARD(s2n_ktls_retrieve_file_descriptor(conn, ktls_mode, &fd));
+    RESULT_GUARD_POSIX(setsockopt(fd, SOL_TCP, TCP_ULP, S2N_TLS_ULP_NAME, S2N_TLS_ULP_NAME_SIZE));
+
+    /* TODO configure keys */
+
+    return S2N_RESULT_OK;
+}
+
+/*
+ * kTLS has been enabled on the socket. Errors are likely to be fatal and
+ * unrecoverable.
+ */
+S2N_RESULT s2n_ktls_configure_connection(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+
+    /* TODO mark kTLS enabled on the connection */
+
+    return S2N_RESULT_OK;
+}
+
+/*
+ * Since kTLS is an optimization, it is possible to continue operation
+ * by using userspace TLS if kTLS is not supported. Upon successfully
+ * enabling kTLS, we set connection->ktls_send_enabled (and recv) to true.
+ *
+ * For this reason we categorize kTLS errors into recoverable and
+ * un-recoverable and handle them appropriately:
+ *
+ * - Errors related to the socket configuration are considered recoverable
+ *   since kTLS related `setsockopt` operations are non-destructive.
+ *
+ * - Errors related to connection configuration are considered
+ *   un-recoverable since we attempt to modify s2n_connection state.
+ */
+int s2n_ktls_enable(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
```

Comment:
Are we communicating what we need to communicate back to the customer? I'm guessing the flow you're expecting is that if s2n_ktls_enable succeeds, then the customer needs to call something like s2n_ktls_is_enabled() to determine whether the connection was actually enabled?

It seems a little strange that a method called s2n_ktls_enable can succeed while not enabling ktls :\

If you must keep the current behavior, I'd explicitly throw one of two ktls errors, one recoverable and one not, and only return success if the recoverable one was set. That will give you better control over what's recoverable and what's not.

Reply:
We are not returning specific errors if kTLS enable fails.